[FLINK-12233][hive] Support table related operations in HiveCatalog

bowenli86 opened the pull request:
## What is the purpose of the change

This PR introduced HiveCatalogTable and implemented table related catalog APIs in HiveCatalog.

## Brief change log

- Created a basic `HiveCatalogTable`, and we can enrich this class later
- Implemented table related catalog APIs in `HiveCatalog`
- Extracted common logic between `HiveCatalog` and `GenericHiveMetastoreCatalog` into `HiveCatalogBase`
- Moved a few util classes from package `org.apache.flink.table.catalog.hive` to `org.apache.flink.table.catalog.hive.util`


## Verifying this change

Reuse tests in CatalogTestBase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)


This PR depends on https://github.com/apache/flink/pull/8339

bowenli86 commented:
cc @xuefuz @KurtYoung @JingsongLi

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:
```diff
@@ -85,49 +86,39 @@ public void alterDatabase(String name, CatalogDatabase newDatabase, boolean igno
-		throw new UnsupportedOperationException();
+	public void alterTable(ObjectPath tablePath, CatalogBaseTable newTable, boolean ignoreIfNotExists)
+			throws TableNotExistException, CatalogException {
+		validateHiveCatalogTable(newTable);
```
Wouldn't we want to do similar validation for GenericHiveMetastoreCatalog?

xuefuz replied:
Similar to createTable(), I don't see we need to provide implementations for alterTable() in each subclass because I don't really see much difference. If validation is different, then we can define a new interface in the base and have each subclass implement it.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java:
```diff
@@ -197,4 +204,140 @@ protected void alterHiveDatabase(String name, Database newHiveDatabase, boolean
+		} else {
+			// IMetastoreClient.alter_table() requires the table to have a valid location, which it doesn't in this case
+			// Thus we have to translate alterTable() into (dropTable() + createTable())
+			dropTable(tablePath, false);
```
I think we needs to be very careful here. dropTable() might also drop the data, but alterTable() really is supposed to change the metadata only.

We should probably solve the location problem rather than taking this workaround.

xuefuz replied:
We also need to note that view and table are sharing this call, so it probably doesn't make sense to change a view to a regular table and vise versa.

bowenli86 replied:
good catch, fixed

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogTable.java:
```diff
@@ -0,0 +1,111 @@
+	@Override
+	public CatalogBaseTable copy() {
+		return new HiveCatalogTable(
+			tableSchema, new ArrayList<>(partitionKeys), new HashMap<>(properties), comment);
```
Don't we need to copy tableSchema?

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveTableConfig.java:
```diff
@@ -19,11 +19,15 @@
 public class HiveTableConfig {
 
-	// Description of the Flink table
+	// -------------------
```
Do the three lines (26-28) provide any additional info above the class name here?

bowenli86 replied:
removed

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java:
```diff
@@ -0,0 +1,148 @@
+	private HiveCatalogUtil() {
+	}
+
+	// ------ Utils ------
```
Nit: The class name is called "utils", and is this line still necessary?

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:
```diff
@@ -85,49 +86,39 @@ public void alterDatabase(String name, CatalogDatabase newDatabase, boolean igno
-	}
-
 	@Override
 	public void createTable(ObjectPath tablePath, CatalogBaseTable table, boolean ignoreIfExists)
```
I think we can keep the implementation only in HiveCatalogBase and define a new API called createHiveTable(). Each subcalss only needs to implement createHiveTable() (because they may create in different ways). I don't see why we need HiveCatalogUtil and related classes for this.

This might be cleaner.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:
```diff
@@ -85,49 +86,39 @@ public void alterDatabase(String name, CatalogDatabase newDatabase, boolean igno
 	@Override
-	public CatalogBaseTable getTable(ObjectPath objectPath) throws TableNotExistException, CatalogException {
-		throw new UnsupportedOperationException();
+	public CatalogBaseTable getTable(ObjectPath tablePath)
```
Similar comment here. I don't see we need getTable() here. getTable() is the same for both catalogs, and the only diff is the translation, which can be captured in a new API.

bowenli86 commented:
Thanks @xuefuz . The reason I use util methods is because as the logic complexity grows, the size of two catalog classes can be quite big. I'm fine to remove util methods and keep shared logic in HiveCatalogBase and special logic of each catalog in its own class now, and we can reevaluate whenever necessary.

Looking at each commit would probably be easier. To summarize the changes:

- moved shared logic to HiveCatalogBase
- added a few interfaces in HiveCatalogBase for subclasses to implement
- updated test util and a few tests to cater to above changes

bowenli86 commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/GenericHiveMetastoreCatalog.java:
```diff
@@ -42,23 +46,35 @@
+	// as Hive metastore has its own properties created upon table creation and migration between different versions of metastore.
+	private static final String FLINK_PROPERTY_PREFIX = "flink.";
+
+	// Flink tables should be stored as 'external' tables in Hive metastore
```
You are right, I verified it locally. I think using Hive managed tables would be fine here.

bowenli86 commented:
@xuefuz @lirui-apache thanks for your review. please take another look.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/GenericHiveMetastoreCatalog.java:
```diff
@@ -95,144 +106,103 @@ public void alterDatabase(String name, CatalogDatabase newDatabase, boolean igno
-				String.format("Failed to drop table %s", tablePath.getFullName()), e);
+	public void validateCatalogBaseTable(CatalogBaseTable table) throws IllegalArgumentException {
+		// TODO: invalidate HiveCatalogView
+		if (table instanceof HiveCatalogTable) {
```
I'm wondering if we should specify what we are expecting and throw exceptions if condition isn't met.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/GenericHiveMetastoreCatalog.java:
```diff
@@ -95,144 +106,103 @@ public void alterDatabase(String name, CatalogDatabase newDatabase, boolean igno
-		return GenericHiveMetastoreCatalogUtil.createCatalogTable(hiveTable);
-	}
+		// Table properties
+		hiveTable.setParameters(buildFlinkProperties(properties));
```
Nit: remove the leading spaces.

bowenli86 replied:
seems to be a github presentation issue, the code doesn't have spaces

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/GenericHiveMetastoreCatalog.java:
```diff
@@ -363,4 +333,23 @@ public CatalogColumnStatistics getPartitionColumnStatistics(ObjectPath tablePath
+	/**
+	 * Add a prefix to Flink-created properties to distinguish them from Hive-created properties.
+	 */
+	private Map<String, String> buildFlinkProperties(Map<String, String> properties) {
```
Maybe we can use words "mask" and "unmask" instead of "build" and "retrieve".

bowenli86 replied:
I feel "mask" is a good name. Not sure about "unmask" because the retrieval not only unmasks flink properties but also filters out hive properties.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java:
```diff
@@ -89,6 +101,34 @@ private static IMetaStoreClient getMetastoreClient(HiveConf hiveConf) {
+	 * @param catalogBaseTable the base table to be validated
+	 * @throws IllegalArgumentException thrown if the input base table is invalid.
+	 */
+	public abstract void validateCatalogBaseTable(CatalogBaseTable catalogBaseTable)
```
Use "protected" if we don't expect external callers.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java:
```diff
@@ -89,6 +101,34 @@ private static IMetaStoreClient getMetastoreClient(HiveConf hiveConf) {
+	 * @param hiveTable a Hive table
+	 * @return a CatalogBaseTable
+	 */
+	public abstract CatalogBaseTable createCatalogTable(Table hiveTable);
```
Can we use word "convertTo" instead of "create"?

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java:
```diff
@@ -89,6 +101,34 @@ private static IMetaStoreClient getMetastoreClient(HiveConf hiveConf) {
+	 * @param table a CatalogBaseTable
+	 * @return a Hive table
+	 */
+	public abstract Table createHiveTable(ObjectPath tablePath, CatalogBaseTable table);
```
Same as above.

bowenli86 replied:
I feel "create" is better because it has more complexity than just simply convert one object to another and keep naming convention consistent. Also feels a bit weird to "convert" two things into one.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java:
```diff
@@ -197,4 +237,193 @@ protected void alterHiveDatabase(String name, Database newHiveDatabase, boolean
+	@Override
+	public void alterTable(ObjectPath tablePath, CatalogBaseTable newCatalogTable, boolean ignoreIfNotExists)
+			throws TableNotExistException, CatalogException {
+		try {
```
Do we need to validate newCatalogTable?

xuefuz replied:
Also, how do we ensure no alter for table to view and vise versa?

bowenli86 replied:
In memory catalog also doesn't check that currently. We should solve that in "FLINK-12452: alterTable() in all catalogs should ensure existing base table and the new one are of the same type"

xuefuz replied:
Yeah. While inmemory is currently doing that, it's achievable, which might not be the case for Hive. This is no consistency required among external catalogs. Nevertheless, I think we should if check if Hive allows it. It's fine to do this in a separate JIRA.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java:
```diff
@@ -197,4 +237,193 @@ protected void alterHiveDatabase(String name, Database newHiveDatabase, boolean
+	/**
+	 * Create a Flink's TableSchema from Hive table's columns and partition keys.
+	 */
+	protected TableSchema createTableSchema(List<FieldSchema> cols, List<FieldSchema> partitionKeys) {
```
This also seems to be conversions. Can we also use word "convertTo" instead of "create"? Same for the next method below.

bowenli86 replied:
ditto

xuefuz replied:
static?

bowenli86 commented:
> Thanks for refactoring the code. It looks much cleaner now.
> 
> However, I still noticed there are still a lot of commonality between the two implementations of createHiveTable() and createCatalogTable(). Maybe we can share more code between them and reduce the code duplication.

@xuefuz they currently look similar mainly because HiveCatalog's implementation is only preliminary at the moment as mentioned in comment. As we enrich HiveCatalog, I expect they will diverge a lot. It would be better to try to unify their common parts in the end.

xuefuz commented:
ok. sounds good. thanks.

bowenli86 commented:
@xuefuz @lirui-apache thanks for your review! The build failure is unrelated (python api issue). If there's no more comment, I will merge tonight

bowenli86 commented:
Rebased to latest master to resolve conflicts. Will wait for a green build before proceeding

xuefuz commented:
> @xuefuz @lirui-apache thanks for your review! The build failure is unrelated (python api issue). If there's no more comment, I will merge tonight
I think I will go thru one more time on this. Thanks.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/GenericHiveMetastoreCatalog.java:
```diff
@@ -363,4 +332,23 @@ public CatalogColumnStatistics getPartitionColumnStatistics(ObjectPath tablePath
+	/**
+	 * Filter out Hive-created properties, and return Flink-created properties.
+	 */
+	private Map<String, String> retrieveFlinkProperties(Map<String, String> hiveTableParams) {
```
static?

bowenli86 replied:
What's the key point of making it static given no other class uses it?

xuefuz replied:
The difference is that a static method doesn't require an instance of the class in order to invoke it. For example, if another static method of this class needs to call this method, without this one defined as static, one has to create an instance of the class to invoke the method. The general principle is that if a method doesn't access any of the object variables/methods, make it static.

bowenli86 replied:
make sense to me. Fixed

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/GenericHiveMetastoreCatalog.java:
```diff
@@ -363,4 +332,23 @@ public CatalogColumnStatistics getPartitionColumnStatistics(ObjectPath tablePath
+	/**
+	 * Add a prefix to Flink-created properties to distinguish them from Hive-created properties.
+	 */
+	private Map<String, String> maskFlinkProperties(Map<String, String> properties) {
```
Static?

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java:
```diff
@@ -89,6 +101,34 @@ private static IMetaStoreClient getMetastoreClient(HiveConf hiveConf) {
+	 * @param catalogBaseTable the base table to be validated
+	 * @throws IllegalArgumentException thrown if the input base table is invalid.
+	 */
+	protected abstract void validateCatalogBaseTable(CatalogBaseTable catalogBaseTable)
```
Even though this is not a public API, I feel it should be more formal. Specifically, we should probably can let this return true or false for validation result, rather than relying on throwing IllegalArgumentException. Secondly, validation can be more more complicated than just check the instance type. At this level, it shouldn't care what kind of validation the sub classes might do.

I think it's acceptable if the base class doesn't care if the sub class does validation or how it does it. So it's okay if we remove this and let each subclass do its validation when create/alter table. This way might be even cleaner.

bowenli86 replied:
This would go against the previous suggestion you made:

> Similar to createTable(), I don't see we need to provide implementations for alterTable() in each subclass because I don't really see much difference. If validation is different, then we can define a new interface in the base and have each subclass implement it.

xuefuz replied:
I'm not backtracking what I said earlier. It's fine to defined an API, but the API needs to be well-thought, which doesn't seem to be the case here. That's why I said we can forgo the API. If we think an API is the way to go, then let's define the API better and parent class shouldn't define what validation a subclass might do. Personally, I don't feel the API is sufficiently sound, but feel free to get 2nd opinion though.

bowenli86 replied:
Make sense. Though IllegalArgumentException can be thrown when any part of the input table is invalid, a better signature may just be throwing CatalogException to conform with that of other existing catalog APIs, since we've defined CatalogException's use case as "in case of any runtime exception".

I don't think returning boolean is a good signature as the check can be fairly complicated and the caller won't know which part is invalid. Returning boolean would be a good option for a single criteria check like "bool isTable()".

xuefuz replied:
If you're concerned about giving user precise, useful information, throwing an exception isn't the best choice. For that, logging is probably better. Neverthelesss, CatalogException certainly sounds better, though it doesn't change the nature of the problem. Let's do this way for now.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java:
```diff
@@ -197,4 +237,196 @@ protected void alterHiveDatabase(String name, Database newHiveDatabase, boolean
+	/**
+	 * Create Hive columns from Flink TableSchema.
+	 */
+	protected List<FieldSchema> createHiveColumns(TableSchema schema) {
```
static?

bowenli86 replied:
I think this is fine as it's used and only used by its two subclasses, not any other classes. Later on we may be able to abstract column and partition creation from subclasses to base class, and this would save us from debating whether it's should be base class or a util class back and forth.

xuefuz replied:
It probably doesn't matter much if one develops both parent class and child classes. Image the base class is defined by one different person, who probably doesn't have much idea what the child class needs. Putting the two utility methods here suggests the parent class needs to envision what's needed by the child classes, which doesn't meet the criteria of separation of concerns. I think moving the two methods to static and out of the parent class would make the code cleaner.

bowenli86 commented:
@xuefuz can you take another look?

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java:
```diff
@@ -89,6 +96,34 @@ private static IMetaStoreClient getMetastoreClient(HiveConf hiveConf) {
+	 * @param hiveTable a Hive table
+	 * @return a CatalogBaseTable
+	 */
+	protected abstract CatalogBaseTable createCatalogBaseTable(Table hiveTable);
```
I think we can just name the method createCatalogTable() which sounds more general, and as a direct counterpart of createHiveTable(). However, this is very minor though.

bowenli86 replied:
This method will return either a CatalogTable or CatalogView, that's why it's named "createCatalogBaseTable() " and "createCatalogTable()" may be a bit misleading

bowenli86 commented:
merging...
